### ISSUE TITLE
User friendly alternative to swiper-avy

### ISSUE THREAD

**Issue by agzam** [abo-abo/swiper#2619]
I use swiper all the time and most of the time I quickly need to jump to the relevant line in the search results. Swiper-avy works fine, but no matter how many times I use it, I can't get used to it, for me it always faster to navigate to the line, and doing swiper-avy way, feels like too much ceremony.

Is is possible to somehow make it so you can jump to the line by typing the exact line number? Or maybe (by default) would let you type something like `foo:31` and it would narrow results to line numbers that contain 31.

**Comment by mohkale** [comment]
`swiper-include-line-number-in-search` is a variable which tells swiper to include line numbers in the search. It doesn't seem to work with `swiper-isearch` but otherwise should fix the second issue.

Aside from that I'm still not quite sure what you're trying to do here. So, say you're at some line in the ivy interface, line 6 and you see that 4 lines down there's the line you want to be at at line 10. Do you want a function which jumps to line 10 (and then you can call `ivy-done` or `ivy-alt-done` to exit)? or a function which moves 4 lines down? 

If it's the latter case you can use the `universal-argument` command and then `ivy-next-line`. So I would type `C-u 4 C-n` (NOTE 4 is optional here because `C-u` is equivalent to 4). If you want to move 3 lines down, `C-u 3 C-n`.

If it's the first case you can use the following function.

```elisp
(defun ivy-goto-line (&optional line)
  "goto ivy candidate at LINE."
  (interactive "P")
  (if line
      (setq line (prefix-numeric-value line))
    (setq line (read-number "line: ")))

  (when (or (<= line 0)
            (> line ivy--length))
    (user-error "line out of range %s:%s" line ivy--length))

  (ivy-set-index (- line 1)))
```

bind it to a key, eg. `M-g`:

```elisp
(general-define-key
 :keymaps 'ivy-minibuffer-map
 "M-g" 'ivy-goto-line)
```

and then hit `M-g` in any ivy interface and you'll be prompted for which candidate you'd like to goto. so `M-g 5 RET` will take you to the 5th candidate. You can also use the universal-argument here `C-u 5 M-g` also works.

Personally I prefer avy to the `ivy-goto-line` and I can live with hitting `ivy-next-line` 3 times instead of using `C-u 3 C-n`. But that's upto personal taste.

**Comment by agzam** [comment]
I am so confused about all what you're saying. I simply want to quickly jump to the line in the document. Here's an example: I have a document that has a few occurrences of word "type". I open swiper and start typing...

![image](https://user-images.githubusercontent.com/1091022/85821612-dd27b600-b72d-11ea-9a11-23aa36873857.png)

A few candidates show up, now I want to quickly jump to the line 17. What options do I currently have? 
- Either navigate to that line using arrow-keys or C-n, or mentally calculate how many lines below the candidate is and use universal argument as you suggested, or
- Use swiper-avy. In my case it requires pressing `C-'` 

![image](https://user-images.githubusercontent.com/1091022/85822007-b28a2d00-b72e-11ea-9aab-c4dac3d9c76c.png)

and now I have to type `lf` to go to the line. 

But why can't I simply type `17` and jump to that line? I can clearly see, the thing I'm looking for is on line 17. My mind doesn't want to think about relative numbers, universal arguments, additional alpha-characters, etc. I see 17. I need to find a way to either jump to 17 or reduce number of candidates to line numbers that have 17 in them (for example if there is also a match in the document on line 170 or 117 or 1017, etc.)

Is that possible?

**Comment by mohkale** [comment]
@agzam yeah, I don't think it is possible. At least not with an avy like interface. How do you distinguish:

```
8   hello world
80 even more hello world
```

## Alternative Solutions

Unless you're willing to also press enter after you've chosen a line to go to. Or are open to padding leading line numbers with 0.

```
08 hello world
80 even more hello world
```

neither of which are very intuitive solutions.

## Avy Smmary

To the best of my understanding, avy works on a tree like this:

```elisp
((97 leaf #<marker at 711 in *scratch*> . #<window 23 on *scratch*>)
 (115 leaf #<marker at 734 in *scratch*> . #<window 23 on *scratch*>)
 (100 leaf #<marker at 752 in *scratch*> . #<window 23 on *scratch*>)
 (102 leaf #<marker at 810 in *scratch*> . #<window 23 on *scratch*>)
 (103 leaf #<marker at 859 in *scratch*> . #<window 23 on *scratch*>)
 (104 leaf #<marker at 912 in *scratch*> . #<window 23 on *scratch*>)
 (106 leaf #<marker at 936 in *scratch*> . #<window 23 on *scratch*>)
 (107 leaf #<marker at 940 in *scratch*> . #<window 23 on *scratch*>)
 (108
  (97 leaf #<marker at 941 in *scratch*> . #<window 23 on *scratch*>)
  (115 leaf #<marker at 968 in *scratch*> . #<window 23 on *scratch*>)
  (100 leaf #<marker at 984 in *scratch*> . #<window 23 on *scratch*>)
  (102 leaf #<marker at 988 in *scratch*> . #<window 23 on *scratch*>)
  (103 leaf #<marker at 989 in *scratch*> . #<window 23 on *scratch*>)
  (104 leaf #<marker at 1009 in *scratch*> . #<window 23 on *scratch*>)
  (106 leaf #<marker at 1048 in *scratch*> . #<window 23 on *scratch*>)
  (107 leaf #<marker at 1075 in *scratch*> . #<window 23 on *scratch*>)))
```

97, 115, 100 etc. are ascii codes for the key you need to press to enter that part of the tree. My earlier example would be:

```elisp
((56 leaf #<marker at 711 in *scratch*> . #<window 23 on *scratch*>) ;; go to line 8
 (56 ;; enter the tree of numbers with a leading 8
  (48 leaf #<marker at 941 in *scratch*> . #<window 23 on *scratch*>))) ;; go to line 80
```

when the user enters 8, how do you know they want to go down the subtree that can take them to line 80, and not immediately exit with 8?

### require user to press enter
The only way this could work is if the user is made to press enter (or some other key) after they've inputted the appropriate line number. The tree would now look like this.

```elisp
((56 ;; enter the tree of numbers with a leading 8
  (13 leaf #<marker at 711 in *scratch*> . #<window 23 on *scratch*>)   ;; go to line 8
  (48 leaf #<marker at 941 in *scratch*> . #<window 23 on *scratch*>))) ;; go to line 80
```

13 is the keycode for `C-m` which is also the return key in a lot of shells.

thoughts?
